src/index/mod.rs

```diff
@@ -137,7 +137,7 @@ impl PartialOrd<Idx> for Idx {
 /// to understand which index algorithms support which operations.
 ///
 /// [`FlatIndex`]: flat/struct.FlatIndex.html
-pub trait Index {
+pub trait Index<Data = f32, Radius = f32> {
```

Comment:
Well, this is an interesting idea, but we may have a problem here. To have a generic Index trait that could cover both binary indexes and real-space vector indexes we need to ensure that they always have the same set of methods. And right now it appears that `reconstruct` and `reconstruct_n` are only available in the latter. It may have been overlooked, but `reconstruct` and `reconstruct_n` for binary indexes are yet to be added to the C API.

In the meantime, we may have to use binary indexes as a separate trait altogether. Maybe once the C API is updated we can think about merging them into a single trait, if in the future we find that to be viable.

Reply:
OK, I must have overlooked something around the  `_reconstruct` methods myself. They seem to be there and ready for use. :shrug: 

> Besides, any future API changes unique to a binary index (or a floating-point index) can be introduced as extension traits for specializations, no?

It would probably work with the assumption that indexes are powered by the same native implementation. OTOH the rest of the design admits the possibility of someone creating their own index implementation in Rust and using that as a drop-in replacement With the extension trait pattern we cannot do this because they are tied to a single implementation. Separate traits or supertraits would be fine though, in faiss-rs we have this pattern in `ConcurrentIndex`.

Reply:
Sweet. Separate traits make sense to me.